newIDE/app/src/ProjectsStorage/ResourceMover/LocalResourceMover.js

```diff
@@ -7,13 +7,178 @@ import {
+import { processByChunk } from '../../Utils/ProcessByChunk';
+import { readLocalFileToFile } from '../../Utils/LocalFileUploader';
 const path = optionalRequire('path');
 
 const gd: libGDevelop = global.gd;
 
+const moveNothing = async () => {
+  return {
+    erroredResources: [],
+  };
+};
+
+const isURL = (filename: string) => {
+  return (
+    filename.startsWith('http://') ||
+    filename.startsWith('https://') ||
+    filename.startsWith('ftp://') ||
+    filename.startsWith('blob:') ||
+    filename.startsWith('data:')
+  );
+};
+
+const isBlobURL = (filename: string) => {
+  return filename.startsWith('blob:');
+};
+
+type ResourceAndFile = {|
+  resource: gdResource,
+  file: File,
+|};
+
+export const moveAllLocalResourcesToCloudResources = async ({
+  project,
+  authenticatedUser,
+  oldFileMetadata,
+  newFileMetadata,
+  onProgress,
+}: MoveAllProjectResourcesOptions): Promise<MoveAllProjectResourcesResult> => {
+  const result: MoveAllProjectResourcesResult = {
+    erroredResources: [],
+  };
+
+  const newCloudProjectId = newFileMetadata.fileIdentifier;
+
+  const resourcesManager = project.getResourcesManager();
+
+  /**
+   * Find the local resources that must be
+   * uploaded into the new project.
+   */
+  const getResourcesToUpload = (project: gdProject): Array<gdResource> => {
+    const allResourceNames = resourcesManager.getAllResourceNames().toJSArray();
+    return allResourceNames
+      .map(
+        (resourceName: string): ?gdResource => {
+          const resource = resourcesManager.getResource(resourceName);
+          const resourceFile = resource.getFile();
+
+          if (isURL(resourceFile)) {
+            if (checkIfIsGDevelopCloudBucketUrl(resourceFile)) {
+              // URL from a cloud project: this is unlikely and would not work
+              // (as the project is a local project). Still, ignore this.
+              return null;
+            } else if (isBlobURL(resourceFile)) {
+              result.erroredResources.push({
+                resourceName: resource.getName(),
+                error: new Error('Unsupported blob URL.'),
+              });
+              return null;
+            } else {
+              // Public URL resource: nothing to do.
+              return null;
+            }
+          } else {
+            // Local resource to be downloaded.
+            return resource;
+          }
+        }
+      )
+      .filter(Boolean);
+  };
+
+  const allResourcesToUpload = getResourcesToUpload(project);
+
+  const projectPath = path.dirname(oldFileMetadata.fileIdentifier);
+
+  let alreadyDoneCount = 0;
+  // Read all files as Files.
+  await processByChunk(allResourcesToUpload, {
+    transformItem: async (resource): Promise<ResourceAndFile | null> => {
+      const resourceAbsolutePath = path.resolve(
+        projectPath,
+        resource.getFile()
+      );
+
+      try {
+        return {
+          resource,
+          file: await readLocalFileToFile(resourceAbsolutePath),
```

Comment:
i'm not at ease with awaits inside objects like this. Can we move it outside for readability?

Reply:
Done 👍